lib: os: PoT utilities, hash function, hash map (hash table)

cfriedt opened the pull request:
Note: **Compliance failure is a false positive**

* Add a few Power-of-Two macros
  * `IS_POWER_OF_TWO()`
  * `LOG2()`
  * `LOG2CEIL()` for `ceil(log2(x))`
  * `NHPOT()` for `2^ceil(log2(x))`
  * compile-time constant and runtime efficient
* Add tests for Power-of-Two macros
* Add a flexible means of specifying a system-wide 32-bit hash function
  * Reference implementation 1: [Daniel J. Bernstein's algorithm (djb2)](https://theartincode.stanis.me/008-djb2/)
  * Reference implementation 2: [Murmur3](https://en.wikipedia.org/wiki/MurmurHash)
  * Add tests to verify uniformity ([Kolmogorov Smirnov](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test))
* Add a flexible means of specifying a system-wide Hashmap implementation
  * Allocator / hash function / load factor may be specified flexibly
  * Allocator should follow `realloc()` semantics, with `free()`-like behaviour when size 0 is specified
  * Implement a [Separate-Chaining](https://en.wikipedia.org/wiki/Hash_table#Separate_chaining) Hashmap
  * Implement an [Open-Addressing](https://en.wikipedia.org/wiki/Hash_table#Open_addressing) Hashmap with Linear Probe
  * Add a C-Wrapper around [`std::unordered_map`](https://en.cppreference.com/w/cpp/container/unordered_map) for benchmarking purposes
  * Performance of `sys_hashmap_oa_lp` on par with `std::unordered_map`
* Add tests to verify all operations of the Hashmap implementations
* Add samples to demonstrate / benchmark different Hashmap implementations

Commit:
lib: os: add hash function specification and implementation

Hash functions are necessary for Hashmaps (a.k.a. Hash Tables).

* Create a flexible hash function interface
* Implementation 1. Murmur3
* Implementation 2. djb2
* Kconfig option for system-wide 32-bit hash

Signed-off-by: Chris Friedt <cfriedt@meta.com>

Commit:
lib: os: add hashmap support

* Add a flexible Hashmap API
* Add a Separate-Chaining Hashmap Implementation
* Add a Open-Addressing Linear Probe Hashmap Implementation
* Add a C-Wrapper for `std::unordered_map` for benchmarking

Signed-off-by: Chris Friedt <cfriedt@meta.com>

Commit:
tests: lib: hashmap: tests for hashmap library

Add tests for public API of `<zephyr/sys/hash_map.h>`

Signed-off-by: Chris Friedt <cfriedt@meta.com>

Commit:
tests: lib: hash_function: tests for sys_hash32

Add a test for `sys_hash32()`. The expectation is that hash
functions should be approximately uniform over a given field.

We can use the Kolmogorov Smirnov test to verify that our
hash function is approximately uniform over a given field.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented on tests/lib/hash_function/src/main.c:
```diff
@@ -0,0 +1,149 @@
+	 *
+	 * E.g. For alpha = 0.05, the estimator is 1.36 / sqrt(n)
+	 *
+	 * However, since we lack sqrt(n), we have to square both sides
```
This can't be right..

cfriedt commented:
I think I'll likely need to add open addressing for performance reasons.

carlocaione commented on include/zephyr/sys/hash_function.h:
```diff
@@ -0,0 +1,80 @@
+ * smaller, fixed-size space. For a given input, a hash function will
+ * consistently generate the same, semi-unique numerical value. Even for
+ * marginally different data, a good hash function will distribute the entropy
+ * almost evenly over all bits in the hashed value. When combined with modulu
```
s/modulu/modulo

carlocaione commented on include/zephyr/sys/hash_function.h:
```diff
@@ -0,0 +1,80 @@
+ *
+ * @return the numeric hash associated with @p str
+ */
+static inline uint32_t sys_hash32(const void *str, size_t n)
+{
+	if (IS_ENABLED(CONFIG_SYS_HASH_FUNC32_CHOICE_DJB2)) {
+		return sys_hash32_djb2(str, n);
+	}
+
+	__ASSERT(0, "No default 32-bit hash. See CONFIG_SYS_HASH_FUNC32_CHOICE");
+
+	return 0;
+}
```
how does this work if the user wants to add its own custom hashing function? Can you extend the test to cover this case?

cfriedt replied:
This inline is mainly for the default, system-wide 32-bit hash function.

If a user wants to use a custom hash function for a Hashmap, then it needs to be a part of the `sys_hashmap.api`. See `SYS_HASHMAP_DEFINE_ADVANCED()`. I could probably document that better.

carlocaione commented on include/zephyr/sys/hash_map_api.h:
```diff
@@ -0,0 +1,194 @@
+ * @param get Retrieve a the value associated with a given key from the Hashmap
+ */
+struct sys_hashmap_api {
+	sys_hash_func32_t func;
```
nit: can we call this `hash_func` or `hash`?

cfriedt replied:
Sure - I have no major preference. Currently refactoring is not working for me in VS code though, so I'll do this later.

cfriedt commented on include/zephyr/sys/hash_map_api.h:
```diff
@@ -0,0 +1,194 @@
+ * @brief Generic Hashmap iterator interface
+ *
+ * Aside from @a map, and @a has_next, all fields are considered invalid
+ * until @a has_next is executed. Subsequently, @a key and @a value are
```
Need to remove references to has_next here.

cfriedt commented:
Need to also make test permutations for each of the different libc's because the default allocator is `realloc()` and consistent behaviour is required

cfriedt commented:
Just for (perf) kicks, I'm going to "implement" another Hashmap which wraps around `std::unordered_map`. Will attempt to create a custom allocator and hash function too, so that the user-provided hooks are still called to level the playing field a bit more. From there, I'll compare each implementation using a flame graph.

cfriedt commented:
TODO:
* bring in user-specified hash function in CXX Hashmap
* bring in user-specified allocator in CXX Hashmap
* run all of the above in native-posix under `perf` to generate FlameGraphs

cfriedt commented on include/zephyr/sys/hash_map.h:
```diff
@@ -0,0 +1,353 @@
+ */
+#define SYS_HASHMAP_DEFINE_STATIC_ADVANCED(_name, _api, _config_type, _data_type, _hash_func,      \
+					   _alloc_func, ...)                                       \
+	static const struct _config_type _name##_config = __VA_ARGS__;                             \
```
These should probably be cast as well

Commit:
samples: lib: os: hash_map: add Hashmap sample code

This sample shows how to use `sys_hashmap`.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
The following benchmarks are done with variations of the following command
```
west build -p auto -b native_posix_64 samples/basic/hash_map/ -- -DCONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES=$((2**21)) -DCONFIG_SYS_HASH_MAP_CHOICE_OA_LP=y
```

Separate Chaining
```
[00:00:00.000,000] <inf> hashmap_sample: n_insert: 2097152 n_remove: 2097152 n_replace: 2097152 n_miss: 0 n_error: 0 max_size: 2097152
[00:00:00.000,000] <inf> hashmap_sample: success
```
![sc](https://user-images.githubusercontent.com/983494/210703953-dcd63a3e-c98d-428d-bc0f-f23283a062b9.svg)


Open Addressing / Linear Probe
```
[00:00:00.000,000] <inf> hashmap_sample: n_insert: 2097152 n_remove: 2097152 n_replace: 2097152 n_miss: 0 n_error: 0 max_size: 2097152
[00:00:00.000,000] <inf> hashmap_sample: success
```
![oa_lp](https://user-images.githubusercontent.com/983494/210703682-80b787eb-2abc-4e80-b49b-55f0d95fd2ab.svg)

C++ Wrapper
```
[00:00:00.000,000] <inf> hashmap_sample: n_insert: 2097152 n_remove: 2097152 n_replace: 2097152 n_miss: 0 n_error: 0 max_size: 2097152
[00:00:00.000,000] <inf> hashmap_sample: success
```
![cxx](https://user-images.githubusercontent.com/983494/210704100-a36691a0-452c-4067-80f2-0bc487d31dc9.svg)

cfriedt commented:
> Just for (perf) kicks, I'm going to "implement" another Hashmap which wraps around `std::unordered_map`. Will attempt to create a custom allocator and hash function too, so that the user-provided hooks are still called to level the playing field a bit more.

It was not easy / feasible to add a custom allocator or custom hash function to the C++ wrapper because those items in standard C++ are structures / functors that do not take any state, whereas we do need to associate them with the hashmap.

Realistically though, the C++ version is really just for benchmarking - if using C++, then it's best to simply use the native `std::unordered_map<K,V>`.

peterdelevoryas commented on lib/os/hash_map_oa_lp.c:
```diff
@@ -0,0 +1,326 @@
+	TOMBSTONE,
+};
+
+struct oalp_entry {
+	uint64_t key;
+	uint64_t value;
+	enum bucket_state state;
+};
```
This is probably the most simple way to store the keys, which is fine, but makes me want to try storing the keys and values separately, and encoding the bucket state in the key, and adding another benchmark comparision hehe

cfriedt replied:
Yep - I thought about that as well. It would be a good experiment, because padding the keys / values just seems wrong.

peterdelevoryas commented on lib/os/hash_map_oa_lp.c:
```diff
@@ -0,0 +1,326 @@
+
+	__ASSERT_NO_MSG(n_buckets != 0);
+
+	for (i = 0, j = hash % n_buckets; i < n_buckets; ++i, ++j, j %= n_buckets) {
```
Maybe we should be making sure that n_buckets is a power of two and using `&` instead of modulus to avoid divides? And modulus is not just one divide, it's multiple, right? And this is in the search path.

Probably not a serious consideration in this first version though, maybe better to just make sure it works.

cfriedt replied:
Would make sense. It would need to be enforced that the table size is always a PoT

peterdelevoryas commented on lib/os/hash_map_oa_lp.c:
```diff
@@ -0,0 +1,326 @@
+		/* grow to a suitable initial size */
+		*new_num_buckets = ceiling_fraction(new_size * 100, map->config->load_factor);
+		return true;
+	} else if (grow && ((new_size + data->n_tombstones) * 100) / data->n_buckets >=
+				   map->config->load_factor) {
+		/* grow at a rate of 2x */
+		*new_num_buckets = data->n_buckets << 1;
+		return true;
```
Aha! Suggestion: grow by 1.5x, not 2x.

https://github.com/facebook/folly/blob/main/folly/docs/FBVector.md#memory-handling

cfriedt replied:
Are you thinking about rehashing in place? That seemed a bit more challenging tbh.

peterdelevoryas replied:
Ohh...oh right. Yeah nevermind my comment. I was thinking "Oh I wonder what growth factor we used" and forgot that this isn't a grow-in-place situation. I guess I'm not even sure how rehashing in-place would work, would definitely be challenging.

peterdelevoryas reviewed (dismissed):
Haven't checked everything, but overall looks like a good starting point!

Commit:
tests: unit: test for Power-of-Two utilities

Tests for Power-of-Two (PoT) utilities.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
Heh.. was consistently getting a factor of 8-10x worse performance than CXX, and then I looked at the naive hash function (djb2), which uses 8 single-byte reads to read a `uint64_t` key from memory, and replaced it with this:

```
uint32_t sys_hash32_djb2(const void *str, size_t n)
{
	uint64_t *u64 = (uint64_t *)str;
	return (uint32_t)*u64;
}
```

Now it's about as fast as CXX. Will merge optimizations shortly.. I'll make an identity hash function as well.

cfriedt commented:
Some additional considerations. Zephyr being an RTOS and all, it sometimes makes sense to limit how big things can get. Is there significant demand for something like that with this API?

I.e. within `struct sys_hashmap_config` should there be a `size_t max_size` field?

cfriedt commented:
For reference:

CXX
```
west build -t run -p auto -b native_posix_64 samples/basic/hash_map/ -- -DCONFIG_SYS_HASH_MAP_CHOICE_CXX=y -DCONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES=$((2**23))
time ./build/zephyr/zephyr.elf 
[00:00:00.000,000] <inf> hashmap_sample: n_insert: 8388608 n_remove: 8388608 n_replace: 8388608 n_miss: 0 n_error: 0 max_size: 8388608
[00:00:00.000,000] <inf> hashmap_sample: success

real    0m0.841s
user    0m0.697s
sys     0m0.145s
```

OA_LP:
```
west build -t run -p auto -b native_posix_64 samples/basic/hash_map/ -- -DCONFIG_SYS_HASH_MAP_CHOICE_OA_LP=y -DCONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES=$((2**23)) -DCONFIG_SYS_HASH_FUNC32_CHOICE_IDENTITY=y
time ./build/zephyr/zephyr.elf 
[00:00:00.000,000] <inf> hashmap_sample: n_insert: 8388608 n_remove: 8388608 n_replace: 8388608 n_miss: 0 n_error: 0 max_size: 8388608
[00:00:00.000,000] <inf> hashmap_sample: success

real    0m0.954s
user    0m0.608s
sys     0m0.347s
```

cfriedt commented:
@JordanYates - closed #53583 in favour of just putting everything here, since there is a direct dependency. Feel free to bring that conversation here.

cfriedt commented:
To leave room for others to add some work later, I think it would be great to have
1. [Paul Hsieh's SuperFastHash](https://chromium.googlesource.com/chromium/src/base/+/master/third_party/superfasthash/superfasthash.c)
2. Robin Hood Hashing

added at a later date.

cfriedt commented:
`qemu_cortex_a53` not passing here and there, oddly, even though everything else is. I'll give it a look later.

cfriedt commented:
@JordanYates - it seems clang has an issue with my version of NHPOT. I've incorporated yours in `sys/util.h` and `sys/util_macro.h`. PTAL

cfriedt commented:
Bitten by the `TIMER_RANDOM_GENERATOR` issue. That's what I get for rebasing, I guess! 😆

JordanYates commented on include/zephyr/sys/util.h:
```diff
@@ -546,6 +546,35 @@ char *utf8_lcpy(char *dst, const char *src, size_t n);
 	})
 
+/** @brief Calculate the base-2 logarithm */
+#define LOG2(x) (sizeof(uint64_t) * 8 - __builtin_clzll((uint64_t)(x)-1))
```
What is the behaviour here with smaller types? The majority of Zephyrs targets are 32bit instead of 64bit AFAIK, so it seems likely that the usual use case would be calling this on something <= 32 bits. It seems likely this is quite inefficient.

JordanYates commented on include/zephyr/sys/util.h:
```diff
@@ -546,6 +546,35 @@ char *utf8_lcpy(char *dst, const char *src, size_t n);
 	})
 
+/** @brief Calculate the base-2 logarithm */
+#define LOG2(x) (sizeof(uint64_t) * 8 - __builtin_clzll((uint64_t)(x)-1))
```
You're super explicit with brackets and conditions elsewhere, its probably worth throwing brackets around `sizeof(uint64_t) * 8`

JordanYates replied:
RE the usage of generics in my original code, they do appear to be used in the Zephyr tree already in the cbprint internals: https://github.com/zephyrproject-rtos/zephyr/blob/9aa4cdbb67a45dcadb7c731de8d17bcb6679eedd/include/zephyr/sys/cbprintf_internal.h#L96
https://github.com/zephyrproject-rtos/zephyr/blob/9aa4cdbb67a45dcadb7c731de8d17bcb6679eedd/include/zephyr/sys/cbprintf_internal.h#L187
etc

cfriedt replied:
> You're super explicit with brackets and conditions elsewhere, its probably worth throwing brackets around `sizeof(uint64_t) * 8`

It's recommended to add brackets around any macro parameters (`x` in this case). Otherwise, it's just plain arithmetic.

> RE the usage of generics in my original code, they do appear to be used in the Zephyr tree already in the cbprint internals:

Yeah... maybe it is best to use generics here as well.

JordanYates commented:
> @JordanYates - it seems clang has an issue with my version of NHPOT. I've incorporated yours in `sys/util.h` and `sys/util_macro.h`. PTAL

No problem, could you please throw a reference in the commit message

JordanYates commented:
Interesting timing on this PR haha #53611

cfriedt commented:
> Interesting timing on this PR haha #53611

I vaguely recall reviewing `LOG2CEIL()` when it was added. It could be simplified quite a bit using the approach that you provided as well.

I'll do another respin and will add a reference in the commit message.

oyvindronningstad commented on include/zephyr/sys/util.h:
```diff
@@ -546,6 +546,55 @@ char *utf8_lcpy(char *dst, const char *src, size_t n);
+ *
+ * @returns log2(x)
+ */
+#define LOG2(x)                                                                                    \
```
This does not handle x=0, which is undefined in `__builtin_clz()`. IMO it should either be supported or mentioned in the macro docs.

oyvindronningstad commented on include/zephyr/sys/util.h:
```diff
@@ -546,6 +546,55 @@ char *utf8_lcpy(char *dst, const char *src, size_t n);
+ * @retval the next highest power of two for @p x
+ * @retval zero, if the next higher power of two for @p x would saturate 32-bits
+ */
+#define NHPOT(x) (1UL << (LOG2(x) + !IS_POWER_OF_TWO(x)))
```
```suggestion
#define NHPOT(x) (1UL << LOG2CEIL(x))
```

cfriedt replied:
> I think these can be made more run-time efficient by adding a `LOG2CEIL()` macro:
> LOG2CEIL: `(NUM_BITS(unsigned int) - __builtin_clz(x - 1))`

It looks like there may be an inconsistency for `LOG2CEIL(x)` when `(x == 0)`, but otherwise, seems reasonable. Likely will use Generics again to address the 32-bit / 64-bit problem.

I might move tests to `tests/unit/pot` instead of `tests/unit/nhpot` since there are more now than just the one macro.

cfriedt replied:
The other thing I'm curious about, is why `LOG2CEIL()` was added to the `toolchain.h` header. I guess I'll move it and see if it breaks anything 😬

oyvindronningstad replied:
> The other thing I'm curious about, is why `LOG2CEIL()` was added to the `toolchain.h` header. I guess I'll move it and see if it breaks anything 😬

I agree. I did search around, and I could not see that it was used anywhere.

oyvindronningstad replied:
Here's a test I wrote for it, in case you find use for it:

```
void test_log2ceil(void)
{
	zassert_equal(0, LOG2CEIL(0), NULL);
	zassert_equal(0, LOG2CEIL(1), NULL);
	zassert_equal(1, LOG2CEIL(2), NULL);
	zassert_equal(2, LOG2CEIL(3), NULL);
	zassert_equal(2, LOG2CEIL(4), NULL);
	zassert_equal(3, LOG2CEIL(5), NULL);
	zassert_equal(3, LOG2CEIL(6), NULL);
	zassert_equal(3, LOG2CEIL(7), NULL);
	zassert_equal(3, LOG2CEIL(8), NULL);
	zassert_equal(4, LOG2CEIL(9), NULL);
	zassert_equal(4, LOG2CEIL(10), NULL);
	zassert_equal(4, LOG2CEIL(11), NULL);
	zassert_equal(4, LOG2CEIL(12), NULL);
	zassert_equal(4, LOG2CEIL(13), NULL);
	zassert_equal(4, LOG2CEIL(14), NULL);
	zassert_equal(4, LOG2CEIL(15), NULL);
	zassert_equal(4, LOG2CEIL(16), NULL);
	zassert_equal(5, LOG2CEIL(17), NULL);
	zassert_equal(5, LOG2CEIL(25), NULL);
	zassert_equal(5, LOG2CEIL(32), NULL);
	zassert_equal(6, LOG2CEIL(33), NULL);
	zassert_equal(6, LOG2CEIL(55), NULL);
	zassert_equal(6, LOG2CEIL(64), NULL);

	for (int i = 1; i < 32; i++) {
		zassert_equal(i, LOG2CEIL((1UL << (i - 1)) + 1), NULL);
		zassert_equal(i, LOG2CEIL(1UL << i), NULL);
	}
}
```

oyvindronningstad replied:
> It looks like there may be an inconsistency for `LOG2CEIL(x)` when `(x == 0)`

Yes x=0 (and x=1) must be handled separately.

oyvindronningstad replied:
I thought the usage in ld files is because LOG2CEIL is a builtin in ld: https://sourceware.org/binutils/docs/ld/Builtin-Functions.html#index-LOG2CEIL_0028exp_0029

Unless you tried removing it and ran into trouble.

oyvindronningstad replied:
The LOG2CEIL (now LOG2CEIL_LD) macro was added in d91e4a52de72306b7de6ee1e193120b8dfc75947, as part of https://github.com/zephyrproject-rtos/zephyr/pull/28231 

cc @bbolen

cfriedt replied:
Interesting - thanks for pointing out the link. Let me rework it again.

Commit:
sys: util: Power-of-Two utilities

* remove previous `LOG2CEIL()` from `<zephyr/toolchain/common.h>`
  - `LOG2CEIL()` is not used in-tree by any C or C++ source
  - only used are in `.ld` files
  - however, it is built-in to both GNU and LLVM linkers
* sys: util: provide LOG2() and LOG2CEIL() macros
  - calculate `log(x)` and `ceil(log(x))`, respectively
  - both are compile-time `const` / ok for `constexpr`
  - also runtime efficient with `__builtin_clz()`
  - `LOG2()`, `NUMBITS()`, and `IS_POWER_OF_TWO()` macros
    contributed by @JordanYates
  - `LOG2CEIL()` contributed by @oyvindronningstad
* sys: util: provide Next-Highest Power-of-Two macro `NHPOT()`
  - calculate `next = pow(2, ceil(log(x)/log(2)))`
  - leverages `LOG2CEIL()`

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
Is it just me, or does `g++` kind of flake-out bigtime when presented with C11 Generics? In contrast, `clang++` handles them without any complaints whatsoever.

<img width="619" alt="Screen Shot 2023-01-13 at 2 26 53 PM" src="https://user-images.githubusercontent.com/983494/212402354-ff2b9588-ad6c-4dbd-b8e6-9e7f58383476.png">

https://en.cppreference.com/w/c/language/generic
https://pastebin.com/QgDT1FGt

cfriedt commented:
Just some final comparisons between OA_LP and CXX 

**OA_LP**
```
# build without profiling (to measure speed)
west build -p auto -b native_posix_64 samples/basic/hash_map/ -- -DCONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES=$((2**23)) -DCONFIG_SYS_HASH_MAP_CHOICE_OA_LP=y -DCONFIG_SYS_HASH_FUNC32_CHOICE_IDENTITY=y

# benchmark speed
time ./build/zephyr/zephyr.elf 
CONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES: 8388608
[00:00:00.000,000] <inf> hashmap_sample: n_insert: 8388608 n_remove: 8388608 n_replace: 8388608 n_miss: 0 n_error: 0 max_size: 8388608
[00:00:00.000,000] <inf> hashmap_sample: success

real    0m0.972s
user    0m0.612s
sys     0m0.360s

# build with profiling (to measure relative efficiency)
west build -p auto -b native_posix_64 samples/basic/hash_map/ -- -DCONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES=$((2**23)) -DCONFIG_SYS_HASH_MAP_CHOICE_OA_LP=y -DCONFIG_SYS_HASH_FUNC32_CHOICE_IDENTITY=y -DCONFIG_GPROF=y
# generate flamegraph
sudo ~/.cargo/bin/flamegraph -o ~/Desktop/oa_lp.svg -F 99 -- ./build/zephyr/zephyr.elf 
```

**CXX**
```
# build without profiling (to measure speed)
west build -p auto -b native_posix_64 samples/basic/hash_map/ -- -DCONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES=$((2**23)) -DCONFIG_SYS_HASH_MAP_CHOICE_CXX=y

# benchmark speed
time ./build/zephyr/zephyr.elf
CONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES: 8388608
[00:00:00.000,000] <inf> hashmap_sample: n_insert: 8388608 n_remove: 8388608 n_replace: 8388608 n_miss: 0 n_error: 0 max_size: 8388608
[00:00:00.000,000] <inf> hashmap_sample: success

real    0m0.777s
user    0m0.619s
sys     0m0.158s

# build with profiling (to measure relative efficiency)
west build -p auto -b native_posix_64 samples/basic/hash_map/ -- -DCONFIG_TEST_LIB_HASH_MAP_MAX_ENTRIES=$((2**23)) -DCONFIG_SYS_HASH_MAP_CHOICE_CXX=y -DCONFIG_GPROF=y
# generate flamegraph
sudo ~/.cargo/bin/flamegraph -o ~/Desktop/cxx.svg -F 99 -- ./build/zephyr/zephyr.elf 
```

cfriedt commented:
I would be fine leaving other optimizations to separate PRs. That would include, for example:
- experimenting with separate allocation strategies for OA_LP key-value pairs and metadata
- hash function optimizations (e.g. perhaps devising a separate Kconfig choice for default Hashmap hash function
  - this could, for example, be used to specify a fixed-size key hash_function (similar to the identity)
  - alternatively, SuperfastHash or something along those lines
- exploring different Hashmap algorithms (e.g. Robin Hood hashing)

oyvindronningstad commented on include/zephyr/sys/util.h:
```diff
@@ -26,6 +26,88 @@
+		return (x > 1) * NUM_BITS(unsigned long long) - __builtin_clzll(x - 1);
+	}
+
+	__ASSERT_NO_MSG(sizeof(x) == sizeof(unsigned long long));
```
What does this assert do? It looks to be unreachable, maybe add a comment or message to explain.

cfriedt replied:
Yea, it was to assert against an unreachable piece of code. I ended up using C++ `static_assert()` because Zephyr's `__ASSERT_NO_MSG()` is not declared and I wasn't sure if it made sense to pull in `<zephyr/sys/__assert.h>` here.

oyvindronningstad commented on include/zephyr/sys/util.h:
```diff
@@ -498,6 +580,35 @@ char *utf8_trunc(char *utf8_str);
+ *
+ * @param x An unsigned integral value
+ *
+ * @return ceil(log2(x)) when 1 <= x <= max(type(x)), 0 when x < 1
```
```suggestion
 * @return ceil(log2(x)) when 1 <= x <= max(type(x)), 0 when x <= 1
```

cfriedt replied:
Actually, both are correct.

oyvindronningstad commented on include/zephyr/sys/util.h:
```diff
@@ -26,6 +26,88 @@
+		  : (NUM_BITS(unsigned long long) - __builtin_clzll((x)-1))))
+
+#define __z_nhpot_impl(x)                                                                          \
+	(((x) == 0) + ((x) != 0) * (_Generic((x), char                                             \
```
I might have it wrong, but I think the `((x) == 0) + ((x) != 0) *` part is not needed. The case of x == 0 is handled correctly with `1 << LOG2CEIL(x)` if the intention is that it returns `1`.

cfriedt replied:
Yep - you were right @oyvindronningstad - thanks for catching that 😁

oyvindronningstad replied:
Did you mean to change this? (It's still the same)

oyvindronningstad replied:
@cfriedt Still curious about this ^.

Otherwise LGTM.

cfriedt replied:
I'll double check. It's possible that I might have forgotten to synchronize on a different machine and undid it by accident.

cfriedt replied:
@oyvindronningstad - fixed!

oyvindronningstad commented on include/zephyr/sys/util.h:
```diff
@@ -517,7 +628,7 @@ char *utf8_lcpy(char *dst, const char *src, size_t n);
 
 /** @brief Number of Hz in @p x kHz */
-#define KHZ(x) ((x) * 1000)
+#define KHZ(x) ((x)*1000)
```
This looks wrong, since it doesn't match the spacing in the `MHZ()` macro below.

cfriedt replied:
That's just auto-formatted according to Zephyr's `.clang-format`, but I can revert it since it's unrelated.

oyvindronningstad commented on include/zephyr/sys/util_macro.h:
```diff
@@ -73,6 +73,9 @@ extern "C" {
 #define BIT64_MASK(n) (BIT64(n) - 1ULL)
 
+/** @brief Check if a @p x is a power of two */
+#define IS_POWER_OF_TWO(x) (((x) != 0U) && (((x) & ((x) - 1U)) == 0U))
```
Curious: What was the reason for extracting this into a macro?

cfriedt replied:
So that it can be used at compile time

oyvindronningstad reviewed (commented):
A few comments on the util.h changes.

cfriedt commented:
* rebased on v3.3.0

cfriedt commented:
* removed unneeded `(x == 0) + (x != 0)* ...` section in `__z_nhpot_impl()`

cfriedt commented:
@mbolivar-nordic - please re-ack when you have a minute

cfriedt commented:
* updated tags in `tests/unit/pot/testcase.yaml`

cfriedt commented:
@oyvindronningstad - will you also ack?

cfriedt commented:
@JordanYates - would you ack?

cfriedt commented:
@carlocaione - look reasonable to you now? If so, please ack :-)

marc-hb commented:
As just discovered by @aborisovich , `constexpr` now requires C++14. Is that intentional and documented? I didn't find anything in the commit messages or in the `doc/` folder

```
zephyr/sys/util.h

#ifdef __cplusplus
template <typename T> static inline constexpr int __z_log2_impl(T x)
```

cfriedt commented:
> As just discovered by @aborisovich , `constexpr` now requires C++14. Is that intentional and documented? I didn't find anything in the commit messages or in the `doc/` folder
> 
> ```
> zephyr/sys/util.h
> 
> #ifdef __cplusplus
> template <typename T> static inline constexpr int __z_log2_impl(T x)
> ```

@marc-hb - this is specifically a workaround for `g++` choking badly when encountering C11 Generics (`clang++` does not). It is documented [here](https://github.com/zephyrproject-rtos/zephyr/pull/53405#issuecomment-1382294647). Feel free to make a PR to document it elsewhere if you like, or a PR to fix `g++`, or some other PR that you feel is more appropriate.

marc-hb commented:
@cfriedt I'm not asking about any specific workaround or other detail. 

https://docs.zephyrproject.org/latest/develop/languages/c/index.html clearly states "In summary, it is recommended to use a compiler toolchain that supports at least the C11 standard for developing with Zephyr"

https://docs.zephyrproject.org/latest/develop/languages/cpp/index.html does not seem to mention any specific C++ version right now. If your recent change requires a C++14 toolchain for any C++ use then it is significantly different from the C11 requirement (3 years newer!) and it should be mentioned on that page.

Zephyr has a very complex toolchain abstraction layer to support a broad range of toolchains, so bumping the requirements and excluding a number of old toolchains is a pretty big deal. A doc update really looks like a minimum.

> Feel free to make a PR to document it elsewhere if you like,

Sorry but I'm not a C++ expert, I don't know which C++ feature is available in each C++ version and I'm not the one who just bumped some C++ requirement. In fact I'm merely trusting @aborisovich here who noticed because his workspace just started failing somehow.

To make sure there's no confusion: I have no opinion on which C++ features should be used in the Zephyr C++ "core" versus not (others might). On the other hand, if C++ requires a newer toolchain than C then a brief mention on https://docs.zephyrproject.org/latest/develop/languages/cpp/index.html looks like a bare minimum. Next time some advance notice would not hurt either.

cfriedt commented:
> https://docs.zephyrproject.org/latest/develop/languages/c/index.html clearly states "In summary, it is recommended to use a compiler toolchain that supports at least the C11 standard for developing with Zephyr"

If your C compiler is ISO compliant, then anything contained within `#ifdef __cplusplus` should be filtered out by the preprocessor. If that's not the case, you may want to look at a different C compiler.

> https://docs.zephyrproject.org/latest/develop/languages/cpp/index.html does not seem to mention any specific C++ version right now. If your recent change requires a C++14 toolchain for any C++ use then it is significantly different from the C11 requirement (3 years newer!) and it should be mentioned on that page.

I would defer to the C++ maintainer for that. CC @stephanosio.

> Zephyr has a very complex toolchain abstraction layer to support a broad range of toolchains, so bumping the requirements and excluding a number of old toolchains is a pretty big deal. A doc update really looks like a minimum.

Again, feel free to make a PR.

> Sorry but I'm not a C++ expert, I don't know which C++ feature is available in each C++ version and I'm not the one who just bumped some C++ requirement. In fact I'm merely trusting @aborisovich here who noticed because his C++ workspace just started failing somehow.

I am also not a C++ expert. Feel free to file an issue, @aborisovich.

> To make sure there's no confusion: I have no opinion on which C++ hard dependencies the Zephyr C++ "core" should have versus not (others might). On the other hand, if C++ requires a newer toolchain than C then a brief mention on https://docs.zephyrproject.org/latest/develop/languages/cpp/index.html looks like a bare minimum. Next time some advance notice would not hurt either.

I'm sorry for the inconvenience. Thank you for bringing it up.

andyross commented:
Not clear on the details here.  Is the contention that Zephyr headers after this patch aren't going to build with existing C++11 projects?  If so that does indeed seem like something we should fix; we're essentially breaking user apps.  If it's just an issue of constexpr that's probably something we can do with some preprocessor magic, we don't generally expose C++ APIs per se, though I do see a few here.

But then as @cfriedt points out, constexpr itself isn't new.  What's the problem that got introduced?

andyross commented:
OK... I'm coming up to speed.  So the issue here (please correct what I'm getting wrong) is that this PR introduces a new generic API implemented **both** with C11 `_Generic` **and** as a C++ template, selected based on source language.

That does seem like it's likely to break things with oddball toolchains, yeah.  Maybe we want to slow-walk this a little and move the new bits out into their own header, so that `zephyr/sys/util.h` (which is included promiscuously by basically everything) gets to stay compatible for longer?

aborisovich commented:
@cfriedt nobody is pointing fingers, this is feedback from us because if external clients of Zephyr project discover that you suddenly require newer standard than before by compilation errors/warnings, that means you are doing something wrong.
There should be a announcement for this and the code that suddenly requires new C++ standard should be probably wrapped in ifdefs that check if the compiler's standard is newer than required or conditionally compiled sources in CMake checking also compiler capabilities.
As @marc-hb always indicates continuous integration should be continuous. And if it can not be, a meaningful and loud announcement should be provided to Zephyr project clients.
Please do so next time if similar situation would take place.

cfriedt commented:
> @cfriedt nobody is pointing fingers, this is feedback from us because if external clients of Zephyr project discover that you suddenly require newer standard than before by compilation errors/warnings, that means you are doing something wrong.

@aborisovich - respectfully, please file a bug report and report details of your issue rather than continuing here. This PR was thoroughly  approved and merged.

> There should be a announcement for this and the code that suddenly requires new C++

As mentioned above, this is all in line with Zephyr's suggested ISO C and C++ standards (C11, C++11).

> standard should be probably wrapped in ifdefs that check if the compiler's standard is newer than required or conditionally compiled sources in CMake checking also compiler capabilities.

Good, then please make a PR. I take it you're using C++98?

> As @marc-hb always indicates continuous integration should be continuous. And if it can not be, a meaningful and loud announcement should be provided to Zephyr project clients.

Please report an issue with whatever detail you have. Zephyr CI is fairly extensive and this passed without issue, which is enough confidence to merge it.

> Please do so next time if similar situation would take place.

Please tell me @aborisovich - how am I supposed to know that your client will have a problem with their CI system. I'm sorry, I don't have ESP.

This is the definition of finger pointing.

Please refer to the Code of Conduct. Thanks.

https://github.com/zephyrproject-rtos/zephyr/blob/main/CODE_OF_CONDUCT.md

cfriedt commented:
> @cfriedt nobody is pointing fingers, this is feedback from us because if external clients of Zephyr project discover that you suddenly require newer standard than before by compilation errors/warnings, that means you are doing something wrong.

@aborisovich - you seem to be very upset about this.

Let me help you:

Point number 1.

Always frame your critique around the code, not the author. The latter makes you come off as being rude & demanding. "you are doing something wrong" as an example. It's probably better to say, this line of code is causing an issue - better yet, just put that in a bug report.

> There should be a announcement for this 

I'm having a hard time understanding why you think the Zephyr Project should change the publicly facing documentation because you discovered a bug that was specific to you project. Why would we make public announcements when someone can just apply a Bugfix? I can only suspect that it's to be able to point a finger.

Point Number 2

Bugs are a fact of life here. If you are not basing your product from a tagged release, then it is an **inevitability** that your code will break. If you are concerned with stability, use a stable release.

> the code that suddenly requires new C++ standard should be probably wrapped in ifdefs that check if the compiler's standard is newer than required or conditionally compiled sources in CMake checking also compiler capabilities.

Point Number 3

Good - you have solved your own problem. For the Nth time, please, file a PR. An issue, sure, if you like.

> As @marc-hb always indicates continuous integration should be continuous. And if it cannot be, a meaningful and loud announcement should be provided to Zephyr project clients

Lol.. no, my friend.

Point number 4

Sorry, but you seem to believe that just because we have CI it's a guarantee that there will never be a single obscure bug due to some wild permutation of config variables.

That's not how this works. If you believe otherwise, I welcome you to read the Apache 2.0 License.

> Please do so next time if similar situation would take place.

This is exactly my point. Stop making demands. Stop telling me what to do. I do not work for you, and I pray I never do. Demanding that someone else fix a bunch of things for you is the wrong approach here.

Had you asked politely in the first place instead of assuming some authoritative role here, I would probably have made the 2-line change for you, which you yourself had described above.

Given all of the above, it makes me infinitely less likely to ever solve a problem for you.

So please, file a PR.

CC @carlescufi

aborisovich commented:
@cfriedt this is not about you. I've submitted the bug to draw attention to the matter so we start to track standard conformance better in Zephyr. Please stop being defensive.

jwakely commented:
G++ doesn't "choke badly", you are trying to use a C feature in C++, that's totally invalid C++ code.

jwakely commented:
> `constexpr` now requires C++14

`constexpr` requires C++11, but there are restrictions on what you can do in a C++11 constexpr function. C++14 relaxes some of the restrictions, and every later standard relaxes them further. But in the simplest form, `constexpr` only requires C++11.

The biggest restriction in C++11 is that a constexpr function can only have a single `return` statement, so any conditional code must be done using `?:` conditional expressions (or template metaprogramming).

jwakely commented:
> Is it just me, or does `g++` kind of flake-out bigtime when presented with C11 Generics? In contrast, `clang++` handles them without any complaints whatsoever. This could pose a bit of a problem, @JordanYates, @oyvindronningstad .
> 
> I might need to revert to the `XXX64()` macros that I had before.
> 
> Update: used C++ templates with `#ifdef __cplusplus`
> 
> <img alt="Screen Shot 2023-01-13 at 2 26 53 PM" width="619" src="https://user-images.githubusercontent.com/983494/212402354-ff2b9588-ad6c-4dbd-b8e6-9e7f58383476.png">

If the issue is simply wanting type-generic `<math.h>` functions, that has worked out of the box in C++ since the first 1998 standard, simply by using overloading. Every math function such as `cbrt` is overloaded by the C++ standard library's version of `<math.h>`:

```c++
#ifdef __cplusplus
extern "C" {
#endif
float cbrtf(float);
double cbrt(double);
long double (cbrtl(long double);
#ifdef __cplusplus
// Additional overloads for float and long double
inline float cbrt(float f) { return cbrtf(f); }
inline long double cbrt(long double f) { return cbrtl(f); }
} // extern "C"
#endif 
```

You don't need to do anything with C11 generics, and you don't need `constexpr`.

cfriedt commented:
> If the issue is simply wanting type-generic `<math.h>` functions, that has worked out of the box in C++ since the first 1998 standard, simply by using overloading. Every math function such as `cbrt` is overloaded by the C++ standard library's version of `<math.h>`:

@jwakely - That is not the issue encountered here but the simplest possible test case to show that g++ fails to parse `_Generic` while LLVM / `clang` does it with ease.

The actual code in question was for compile-time constant power-of-two macros.
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util.h#L595

> You don't need to do anything with C11 generics, and you don't need `constexpr`.

Actually I do, and I do.

Bug was files (and quickly marked invalid / closed) here
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108965

jwakely commented:
> That is not the issue encountered here but the simplest possible test case to show that g++ fails to parse `_Generic` while LLVM / `clang` does it with ease.

That just means clang has a non-standard extension. It's not a GCC bug that it doesn't support it. It's not in C++.

> Bug was files (and quickly marked invalid / closed) here

Yes, I know - I closed it.

Looking at your `__z_log2_impl` implementation for C++, it should not be `static` (that leads to [ODR](https://en.cppreference.com/w/cpp/language/definition#One_Definition_Rule) violations if called from other inline functions or templates, i.e. undefined behaviour). And `constexpr` implies `inline` so that's redundant.

A valid C++11 implementation would be:

```c++
template <typename T> constexpr int __z_log2_impl(T x)
{
	/* No declaration of __ASSERT is available here */
	static_assert(sizeof(x) <= sizeof(unsigned long long), "unsupported type for LOG2()");

	return sizeof(x) < sizeof(unsigned int) ?
		((x) < 1) * (-1) +
		       ((x) >= 1) * (NUM_BITS(unsigned int) - __builtin_clz(x) - 1)
         : sizeof(x) <= sizeof(unsigned long long) ?
		((x) < 1) * (-1) +
		       ((x) >= 1) * (NUM_BITS(unsigned long long) - __builtin_clzll(x) - 1)
         : -1;
}
```

jwakely commented on include/zephyr/sys/util.h:
```diff
@@ -26,6 +26,90 @@
+#define __z_log2_impl(x)                                                                           \
+	(((x) < 1) * (-1) +                                                                        \
+	 ((x) >= 1) * _Generic((x), char                                                           \
+			       : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), unsigned char    \
```
Is omitting `signed char` intentional?

cfriedt replied:
@jwakely - no... because .. it wasn't omitted?

cfriedt replied:
Thanks for clarifying - patches welcome!

cfriedt commented:
> > That is not the issue encountered here but the simplest possible test case to show that g++ fails to parse `_Generic` while LLVM / `clang` does it with ease.
> 
> That just means clang has a non-standard extension. It's not a GCC bug that it doesn't support it. It's not in C++.

@jwakely Oh, ok. Then should I file an issue as a feature request? I didn't see that option in bugzilla. Would it receive any more attention than what it was given today?

> Looking at your `__z_log2_impl` implementation for C++, it should not be `static` (that leads to [ODR](https://en.cppreference.com/w/cpp/language/definition#One_Definition_Rule) violations if called from other inline functions or templates, i.e. undefined behaviour). And `constexpr` implies `inline` so that's redundant.
> 
> A valid C++11 implementation would be:

Thanks for the tips!

mbolivar-nordic commented:
Is there any chance that this discussion could continue in a new issue so that people who were subscribed to the PR do not continue to get updates that may not be relevant to them?

marc-hb commented:
> Is there any chance that this discussion could continue in a new issue so that people who were subscribed to the PR do not continue to get updates that may not be relevant to them?

@aborisovich filed #55204 two days ago and Github automatically added a link above.  Its description seems very detailed and some discussion has already started there. 

Anyone who feels the need to file yet another, separate issue should of course do so.

@andyross just submitted PR #55290
